sql/hive/src/test/scala/org/apache/spark/sql/catalyst/LogicalPlanToSQLSuite.scala

```diff
@@ -988,9 +988,88 @@ class LogicalPlanToSQLSuite extends SQLBuilderTest with SQLTestUtils {
+        |                  AS x
+        |           INNER JOIN (SELECT `key` AS `gen_attr`, `value` AS `gen_attr`
+        |                       FROM `default`.`parquet_t1`)
+        |                       AS gen_subquery_1
+        |           ON (`gen_attr` = `gen_attr`))
+        |           AS gen_subquery_2)
+        |      AS gen_subquery_3)
+        |      AS x
+      """.stripMargin
+    )
+  }
+
+  test("broadcast hint with rollup") {
+    checkGeneratedSQL(
+      """
+        |SELECT /*+ MAPJOIN(parquet_t1) */
+        |       count(*) as cnt, key%5, grouping_id()
+        |FROM parquet_t1
+        |GROUP BY key % 5 WITH ROLLUP""".stripMargin,
+      """
+        |SELECT `gen_attr` AS `cnt`, `gen_attr` AS `(key % CAST(5 AS BIGINT))`, `gen_attr` AS `grouping_id()`
+        |FROM (SELECT /*+ MAPJOIN(parquet_t1) */ count(1) AS `gen_attr`,
+        |             (`gen_attr` % CAST(5 AS BIGINT)) AS `gen_attr`, grouping_id() AS `gen_attr`
+        |      FROM (SELECT `key` AS `gen_attr`, `value` AS `gen_attr`
+        |            FROM `default`.`parquet_t1`)
+        |            AS gen_subquery_0
+        |      GROUP BY (`gen_attr` % CAST(5 AS BIGINT))
+        |      GROUPING SETS(((`gen_attr` % CAST(5 AS BIGINT))), ()))
+        |      AS gen_subquery_1
+      """.stripMargin)
+
+  }
+
+  test("broadcast hint with grouping sets") {
+    checkGeneratedSQL(
+      s"""
+         |SELECT /*+ MAPJOIN(parquet_t1) */
+         |       count(*) AS cnt, key % 5 AS k1, key - 5 AS k2,grouping_id() AS k3
+         |FROM (SELECT key, key % 2, key - 5 FROM parquet_t1) t GROUP BY key % 5, key - 5
+         |GROUPING SETS (key % 5, key - 5)
+      """.stripMargin,
+      """
+        |SELECT `gen_attr` AS `cnt`, `gen_attr` AS `k1`, `gen_attr` AS `k2`, `gen_attr` AS `k3`
+        |FROM (SELECT count(1) AS `gen_attr`, (`gen_attr` % CAST(5 AS BIGINT)) AS `gen_attr`,
+        |             (`gen_attr` - CAST(5 AS BIGINT)) AS `gen_attr`, grouping_id() AS `gen_attr`
+        |      FROM (SELECT /*+ MAPJOIN(parquet_t1) */ `gen_attr`,
+        |                   (`gen_attr` % CAST(2 AS BIGINT)) AS `gen_attr`,
+        |                   (`gen_attr` - CAST(5 AS BIGINT)) AS `gen_attr`
+        |            FROM (SELECT `key` AS `gen_attr`, `value` AS `gen_attr`
+        |                  FROM `default`.`parquet_t1`)
+        |                  AS gen_subquery_0)
+        |            AS t
+        |      GROUP BY (`gen_attr` % CAST(5 AS BIGINT)), (`gen_attr` - CAST(5 AS BIGINT))
+        |      GROUPING SETS(((`gen_attr` % CAST(5 AS BIGINT))), ((`gen_attr` - CAST(5 AS BIGINT)))))
+        |      AS gen_subquery_1
+      """.stripMargin)
+  }
+
+
   // This will be removed in favor of SPARK-16590.
   private def checkGeneratedSQL(sqlString: String, expectedString: String): Unit = {
-    val generatedSQL = new SQLBuilder(sql(sqlString)).toSQL
+    val df = sql(sqlString)
+    logError("\n" + df.queryExecution.analyzed.treeString)
```

Comment:
After https://github.com/apache/spark/pull/14235, I need to change the testcase style, too.